### PR TITLE
sympy replace inplace

### PR DESCRIPTION
**Pull request by cattabiani**
- ~Only small diffEq and LinEq are supported~ Everything is supported
- Identifies diffEq/linEq/nonLinEq and replace them in-place updating all the
  intermediate variables that need updating
- In case some statements are not matching, it still tries to
  replace them inplace with a greedy approach
- If everything fails, ~the remaining solutions are added at the end~ we throw an error: something probably went wrong

TODO (preferably in this order):

- ~Add documentation~
- ~Add warnings and errors~
- ~Add (a lot of) unit tests~
- ~Clean the code from debugging routines~

~Partial~ Fix: #55

### CODE REVIEW

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (ohm314) [inline src/visitors/sympy_replace_solutions_visitor.cpp:165]: I'm a bit confused, I thought `statement_dependencies` is only used with assignment statements? Should the documentation be updated or is there something I don't understand?

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (ohm314) [inline src/visitors/sympy_replace_solutions_visitor.cpp:176]: Why do you do this after you already get `statement_dependenices` on line 165-166?

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (ohm314) [inline src/visitors/sympy_replace_solutions_visitor.cpp:97]: why do you do this on every level and not only root?

**Review by ohm314 — COMMENTED** [review]
Thanks for tackling this @cattabiani ! Overall I think your solution makes sense. I have a few questions to help me understand better.

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp:50]: no. It assumes that the first node where you call it is the root node. In the root node I need to finally replace stuff, set/check 
the interleaves_counter etc. Since I possibly run the visitor twice from the root node I cannot just set everything in the constructor.

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp]: already removed. Something that I forgot

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp:421]: correct! I will change the name. Yours is better

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp]: imagine this situation:

```
~ y = a + x
~ x + 3*y = 1
```

the solution vector is:
```
x = f(...)
y = g(...)
```

first we pass, matching by value and the SECOND solution is replaced (we match y). Then, we pass greedly and we want to replace that by the first tagged equation (that is the first one). I could replace the tagging vector a set... I need to check this

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:255]: ususally this is used to find out what is the solution statement that assigns that variable. That is different from a binary expression in the original code that assigns the variable. I am afraid this would confuse some. In addition, there is a statement_ member in SolutionSorter and this umap points at that

File: src/visitors/visitor_utils.cpp
Comment (cattabiani) [reply src/visitors/visitor_utils.cpp:241]: There are 2 cases:

a' -> a
F'[0] -> F[0] 

I tried with "to_nmodl" but I get the '. I tried with getting only the var_name but I get F only for the second (not F[0]). Do you know of something better?

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp:370]: yes in case we do not match

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp]: Let's say that I have a statement block inside a statement block (or any other block). If I replace only in the root, while I pass I find in the statementVector the full statement block (which does not match any replacement since the remplacements are done diffEq by diffEquation) and I pass over without replacing anything.

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp:50]: no, we can call it also on the child nodes. However, in the root node I need to set and update a few things that are global. Since I may need to pass twice I sometimes need to reset stuff (but only if I am the root node)

Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp:50]: for example: interleaves_counter

File: src/visitors/visitor_utils.cpp
Comment (cattabiani) [reply src/visitors/visitor_utils.cpp:241]: basically I have 2 cases that I need to take care:
a' -> a
F'[0] -> F[0]

if I get the VAR_NAME node I get only F (instead of F[0]) if i use to_nmodl I get a' instead of a. Do you have a better routine?

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp]: not only for assignments. It was moved into utils adding a little bit of documentation

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp]: after checking I decided that a set is better suited for tags_. I proceeded with the chances and it passes the ctests. Let's see for the CI. Thank you for the remark!

**Comment by cattabiani** [comment]
please retest

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (ohm314) [reply src/visitors/sympy_replace_solutions_visitor.cpp:50]: Yeah, I think we agree. I understood that it is called on the root nodes but not only. The question was more, that necessarily it has to be called first on the root node, that's the entry point. But as I said, this is probably all irrelevant, because it's an internal visitor and we have probably plenty of other visitors that are exactly the same. The assumption is that whoever calls this visitor they need to first call it on the root node, and it will then recursively call on child nodes. I'm going to resolve this, since I don't think it's an actual issue, just wanted to clarify.

File: src/visitors/visitor_utils.cpp
Comment (ohm314) [reply src/visitors/visitor_utils.cpp:241]: OK, after having also discussed in chat. the `'` operator could be also parsed as an operator and treated as such, which would probably clean up such manual string manipulation and make the semantics a bit stronger. But, this is not the scope of this PR, so we leave it out. Could you please create an issue though? Then we can decide later whether we want to implement this or not.

Comment (cattabiani) [reply src/visitors/visitor_utils.cpp:241]: added here: https://github.com/BlueBrain/nmodl/issues/511

**Review by pramodk — COMMENTED** [review]
Few comments, but review is WIP.

File: test/unit/visitor/sympy_solver.cpp
Comment (cattabiani) [reply test/unit/visitor/sympy_solver.cpp:696]: not really because they can be used in the solution that you are substituting. For example, if sympy returns with:

```
x = f(x,y)
y = g(x,y)
```

and you do not use old_x old_y you end up using the new x and the old y for the computation of y but the old x and y for the computation of x since when you do the check you analyze the initial unsolved set of equations. You could analyze both old and new set of equations. However, this is much more complicatede and error-prone than just snapshotting x value at the beginning of the cycle

File: test/unit/visitor/sympy_solver.cpp
Comment (cattabiani) [reply test/unit/visitor/sympy_solver.cpp:1349]: no that should probably be fine. I need to verify

Comment (cattabiani) [reply test/unit/visitor/sympy_solver.cpp:1349]: EDIT: I checked. A warning is raised if the greedy approach is required. If the codes thinks that we need to update the pre-solve statements (ex: `old_x = x`) it throws the error that I was talking about before.

File: src/visitors/visitor_utils.cpp
Comment (cattabiani) [reply src/visitors/visitor_utils.cpp:240]: mmm yes, good. Thx

Comment (cattabiani) [reply src/visitors/visitor_utils.cpp:240]: `ast::Expression` does not have a `get_name` function. I could use `get_var_name` but I would miss the `[0]` for the vectors. Since I did not understand your idea fully, I dunno how to correct the code. What is the node that should be used. Notice that this function should work for:

- diffEqExpression
- linEquation
- nonLinEquation
- binary expression

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:80]: it is just that I am used to that since when I was at cern. Do we have another convention?

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:40]: I added a lot of doc. Check it out!

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:50]: I added a lot of doc. Check it out!

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:284]: I added a lot of doc. Check it out!

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:65]: nitpick: use `\code{.mod}` and ` \endcode` to include NMODL code block

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:249]: `the first activation of the switch` : may be just me that couldn't pick this up.

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:322]: nitpick : In other parts of the code, we are keeping `/brief` in the header file but the `/details` in the implementation .cpp file. This way, you have logic near the implementation and it's easy to refer / scroll to an update?

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:361]: nitpick : this `brief` explanation could be better :)

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:370]: similar

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:410]: nitpick : I would prefer `\code{.mod}` block for doxygen documentation.

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:2]:
```suggestion
 * Copyright (C) 2018-2021 Blue Brain Project
```

**Review by pramodk — COMMENTED** [review]
I went through this today and played with some examples. I wouldn't say I know now everything but playing with examples gave me enough confident and it's doing what is supposed to do!

I have some minor comments / suggestions. Apart from this I think it's in good state.

**Comment by pramodk** [comment]
@cattabiani : in case you will need any practical example, all mod files are stored in the below repository:

```
git clone ssh://kumbhar@bbpcode.epfl.ch/incubator/nmodldb -b sandbox/kumbhar/downloader
cd nmodldb/models/db/   #all mod files under this directory
```

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp:20]: I improved the docs. It is a +1 in case `in_system_ == false` and `is_in_system == true` (a switch). Maybe I can rename `in_system_` -> `old_in_system_` to stress out that it follows `is_in_system`

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp:38]: I opened an issue that reflects the TODOs that are in the code. I leave it to another PR since it is likely that `tmp*` and `old_*` are hard-coded around the code and a change here can generate bugs

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp:115]: not very easy since this error is only for developers. If the user reaches this point there was a bug somewhere before. In any case, I improved it because utimately, yes, I agree it was a shitty error message

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp]: it is less verbose but I call `statement_dependencies` even if the greedy approach is employed. in this way it is more efficent because I call it only when I need it. What do you prefer? (I do not have strong opinions on this)

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp:265]: a root_var is a variable that does not depend on other variables. For example:

```
tmp = a
x = tmp + b
```
a and b are root_var. tmp and x are derived. In other words, a root_var is every var that is not assigned

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:441]: perfect

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:65]: I should have done it everywhere. Thanks!

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:249]: improved the docs

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:410]: done

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:370]: done

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:361]: done

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.cpp:20]: improved the docs

**Review by ohm314 — APPROVED** [review]
Very nice, @cattabiani ! I like how you expanded the documentation and debug logging. I think for me this code looks ready to merge, pending @pramodk 's final OK (since he is running through mod files too 😄 )

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [reply src/visitors/sympy_replace_solutions_visitor.hpp:231]: we have not been using `_` or anything else to distinguish public/private in the codebase.

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:66]:
```suggestion
 * \code{.mod}
```
the reason for this is that with `.mod`, doxygen does NMODL syntax highlighting in the docs!

File: src/visitors/sympy_replace_solutions_visitor.cpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.cpp:284]:
```suggestion
 *  \code{.mod}
```

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:87]:
```suggestion
 * \code{.mod}
```

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:95]:
```suggestion
 * \code{.mod}
```

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:108]:
```suggestion
 * \code{.mod}
```

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:116]:
```suggestion
 * \code{.mod}
```

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (pramodk) [inline src/visitors/sympy_replace_solutions_visitor.hpp:180]:
```suggestion
 * \code{.mod}
```

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:66]: aaaaah. I did not understand

File: src/visitors/sympy_replace_solutions_visitor.hpp
Comment (cattabiani) [reply src/visitors/sympy_replace_solutions_visitor.hpp:231]: do you prefer to remove it or I can leave it there?

Comment (pramodk) [reply src/visitors/sympy_replace_solutions_visitor.hpp:231]: My personal preference is to remove it for consistency of codebase

**Review by pramodk — APPROVED** [review]
LGTM 

Feel free to merge with `squash and nice commit message`.  (I would prefer to remove underscores for private members)

**Review by pramodk — APPROVED** [review]

**Comment by cattabiani** [comment]
please retest

**Comment by cattabiani** [comment]
please retest

**Comment by cattabiani** [comment]
please retest

**Pull request merged**
